experimental/dds/tree2/src/feature-libraries/editable-tree-2/proxies/proxyNodeCursor.ts

```diff
@@ -0,0 +1,166 @@
+				? (node as Value)
+				: undefined;
+		},
+		type: (node: TNode) => {
+			if (node === undefined) {
+				throw new Error("undefined node");
+			}
+			if (node === null) {
+				return leaf.null.name;
+			}
+
+			if (isFluidHandle(node)) {
+				return leaf.handle.name;
+			}
+
+			switch (typeof node) {
+				case "number":
+					return leaf.number.name;
+				case "string":
+					return leaf.string.name;
+				case "boolean":
+					return leaf.boolean.name;
+				default: {
+					const possibleTypes = getPossibleTypes(
+						context,
+						typeSet,
+						node as ContextuallyTypedNodeData,
+					);
+					assert(
+						possibleTypes.length !== 0,
+						"data is incompatible with all types allowed by the schema",
+					);
+					assert(
+						possibleTypes.length === 1,
+						"data is compatible with more than one type allowed by the schema",
+					);
+					return possibleTypes[0];
+				}
+			}
+		},
+		keysFromNode: (node: TNode) => {
+			switch (typeof node) {
+				case "object":
+					if (node === null) {
+						return [];
+					} else if (Array.isArray(node)) {
+						return node.length === 0 ? [] : [EmptyKey];
+					} else if (node instanceof Map) {
+						const unfilteredKeys = Array.from(node.keys()) as FieldKey[];
+						// Map proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the entry.
+						const filteredKeys = unfilteredKeys.filter(
+							(key) => node.get(key) !== undefined,
+						);
+						return filteredKeys;
+					} else {
+						// Assume record-like object
+						const objectedNode = node as Record<FieldKey, unknown>;
+
+						const unfilteredKeys = Object.keys(node) as FieldKey[];
+						// Object proxies may contain entries with explicit `undefined` values.
+						// We wish to omit these from our representation.
+						// Setting a key's value to `undefined` is equivalent to omitting the key/value pair.
+						const filteredKeys = unfilteredKeys.filter((key) => {
+							const value = objectedNode[key];
+							return value !== undefined;
+						});
+						return filteredKeys;
+					}
+				default:
+					return [];
+			}
+		},
+		getFieldFromNode: (node: TNode, key: FieldKey): readonly TNode[] => {
+			// Object.prototype.hasOwnProperty can return true for strings (ex: with key "0"), so we have to filter them out.
+			// Rather than just special casing strings, we can handle them with an early return for all primitives.
+			if (typeof node !== "object") {
+				return [];
+			}
+
+			if (node === null) {
+				return [];
+			}
+
+			if (Array.isArray(node)) {
+				// TODO: verify this
+				return key === EmptyKey || key === rootFieldKey ? node : [];
+			}
+
+			if (node instanceof Map) {
+				return (node as Map<FieldKey, TNode[]>).get(key) ?? [];
+			}
+
+			if (Object.prototype.hasOwnProperty.call(node, key)) {
+				const field = (node as Record<FieldKey, TNode>)[key];
```

Comment:
Do we need to [exclude explicit undefineds](https://www.typescriptlang.org/play?#code/MYewdgziA2CmB00QHMAUB5ARgK1sALvAA4BOI+5AnkQgBYCGE6A7mAApk0n6XzD3RoqAN4ACAB4AuUQFcwAE1gAzAJZhY80QF8ANKIBE4-QEpjAbiA)?

Reply:
Oh, yeah we probably do.

Reply:
No. We filter out keys with undefined values in `keysFromNode`. So presumably this doesn't get called for those keys. If it does, I believe it's valid to just return undefined.

Reply:
Yeah, you're probably right. We already filter out keys with undefined values in `keysFromNode`, so presumably this doesn't get called. But in the event that it was called, this would be incorrect. Will update.